src/openrct2/world/Map.cpp

```diff
@@ -1539,6 +1548,106 @@ void map_update_tiles()
     }
 }
 
+void MapInvalidateHeightCache()
+{
+    _mapHighestTileHeightLoopPosition = 0;
+}
+
+void MapUpdateHeightCache(bool fullCheck)
+{
+    constexpr int32_t maxTileIndex = MAXIMUM_MAP_SIZE_TECHNICAL * MAXIMUM_MAP_SIZE_TECHNICAL;
+
+    if (_mapHighestTileHeightLoopPosition == maxTileIndex)
+        return;
+
+    int32_t tilesToCheck = maxTileIndex;
+
+    // If we're doing a full check of the map then zero out the loop
+    // position to start at tile zero, checking all tiles before returning.
+    // Otherwise check a limited number of tiles per each time the
+    // function is called.
+    if (fullCheck)
+        _mapHighestTileHeightLoopPosition = 0;
+    else
+        tilesToCheck /= 64;
+
+    for (int32_t i = 0; i <= tilesToCheck; i++)
```

Comment:
I'm not sure the loop condition is correct. Total iterations would be (64 * 1024) + 64.
See https://godbolt.org/z/jsPvdo

Reply:
Your right. That's a bit embarrassing. I confused myself by naming the tile count maxTileIndex, which is wrong. Will update, sorry about that.

Reply:
No worries, happens to the best. Thats what reviews are good for 😉